plugins/kafka/kafka.cpp

```diff
@@ -591,45 +624,41 @@ namespace KafkaPlugin
 
     KafkaStreamedDataset* Consumer::getMessageDataset(IEngineRowAllocator* allocator, __int64 maxRecords)
     {
-        // Make sure we have a valid connection to the Kafka cluster
-        ensureSetup();
+        try
+        {
+            // Make sure we have a valid connection to the Kafka cluster
+            ensureSetup();
 
-        // Start the local read queue
-        RdKafka::ErrorCode startErr = consumerPtr->start(topicPtr, partitionNum, RdKafka::Topic::OFFSET_STORED);
+            // Start the local read queue
+            RdKafka::ErrorCode startErr = consumerPtr->start(topicPtr, partitionNum, RdKafka::Topic::OFFSET_STORED);
 
-        if (startErr == RdKafka::ERR_NO_ERROR)
-        {
-            if (traceLevel > 4)
+            if (startErr == RdKafka::ERR_NO_ERROR)
+            {
+                if (traceLevel > 4)
+                {
+                    DBGLOG("Kafka: Started Consumer for %s:%d @ %s", topic.c_str(), partitionNum, brokers.c_str());
+                }
+            }
+            else
             {
-                DBGLOG("Kafka: Started Consumer for %s:%d @ %s", topic.c_str(), partitionNum, brokers.c_str());
+                throw MakeStringException(-1, "Kafka: Failed to start Consumer read for %s:%d @ %s; error: %d", topic.c_str(), partitionNum, brokers.c_str(), startErr);
             }
         }
-        else
+        catch (...)
         {
-            throw MakeStringException(-1, "Kafka: Failed to start Consumer read for %s:%d @ %s; error: %d", topic.c_str(), partitionNum, brokers.c_str(), startErr);
+            // The current object is allocated on the heap and was supposed to be managed by the
+            // IRowStream subclass we're returning; since we've encounted an error, we need
+            // to self-destruct to avoid memory leaks
+            delete(this);
```

Comment:
This rings some alarm bells.  Better would be to have the catch in the global getMessageDataset() function and delete consumerObjPtr.  (or use std::unique_ptr)

Reply:
Moving to using a catch in the calling function.